src/pybind/mgr/dashboard/controllers/nvmeof.py

```diff
@@ -392,23 +395,148 @@ def list(self, nqn: str):
+            return status
+
+        @Endpoint('GET', '/subsystems')
+        @ReadPermission
+        def list(self):
+            response = []
+            # Get list of subystems
+            subsystems = NVMeoFClient().stub.list_subsystems(
+                NVMeoFClient.pb2.list_subsystems_req()
+            )
+            if subsystems.status != 0:
+                raise DashboardException(
+                    msg=response.error_message,
+                    code=response.status,
+                    http_status_code=NVMeoFClient().NVMeoFError2HTTP.get(response.status, 400),
+                    component="nvmeof",
+                )
+            # Create a new response with initiators information appended
+            for subs in subsystems.subsystems:
+                res = {
+                    'nqn': subs.nqn,
+                    'enable_ha': subs.enable_ha,
+                    'serial_number': subs.serial_number,
+                    'model_number': subs.model_number,
+                    'min_cntlid': subs.min_cntlid,
+                    'max_cntlid': subs.max_cntlid,
+                    'namespace_count': subs.namespace_count,
+                    'max_namespaces': subs.max_namespaces,
+                    'subtype': subs.subtype,
+                    'initiators': [],
+                }
+                # Get list of initiators present on the subsystem
+                hosts = NVMeoFHost().list(subs.nqn)
+                # Decode the bytes object to json
+                json_str = hosts.decode('utf-8')
+                json_obj = json.loads(json_str)
+                # Appending initiators information
+                res['initiators'] = json_obj
+                response.append(res)
+            return response
+
+        @Endpoint('POST', '/subsystem')
+        @empty_response
+        @handle_nvmeof_error
+        @CreatePermission
+        def create(self, nqn: str, enable_ha: bool,
+                   max_namespaces: int = 256, initiators: str = None):
+            subsystem_response = NVMeoFClient().stub.create_subsystem(
+                NVMeoFClient.pb2.create_subsystem_req(
+                    subsystem_nqn=nqn, max_namespaces=max_namespaces, enable_ha=enable_ha
+                )
+            )
+
+            if subsystem_response.status != 0:
+                return subsystem_response
+
+            if initiators is not None:
+                all_host_nqns = initiators.split(',')
+                for host_nqn in all_host_nqns:
+                    add_host_response = NVMeoFClient().stub.add_host(
+                        NVMeoFClient.pb2.add_host_req(subsystem_nqn=nqn, host_nqn=host_nqn)
+                    )
+                    if add_host_response.status != 0:
+                        return add_host_response
+
+            return subsystem_response
+
+        @staticmethod
+        def determine_nqns(res, new_host_nqns):
+            to_delete_nqns = []
+            to_add_nqns = []
```

Comment:
This method is bit risky, as it's trying to do too much: converting an imperative interface to a declarative one. That's why we preferred to use POST and PATCH rather than PUT, and implement REST methods that mimic the CLI, rather than providing a super-high level REST API.

This REST API (nor the nvmeof CLI) doesn't provide any transaction safety guarantees, so if "someone" added/removed another host while this transaction is running (not likely, but not impossible), this method will delete/re-add that or will simply end up with a list of host NQNs different to what was stated in the REST PUT request.

Reply:
The goal is to allow user to add/delete multiple hosts rather than single one(_as done by CLI already_)
If you have any better suggestions to do that let me know, will update.

Reply:
Users can still add/remove multiple hosts... Only that it'll mean more `POST`/`DELETE` requests. Personally, I still think that it's not a good thing that the Dashboard is not using its own REST API.

The current NVMe REST API allows the UI to do anything (it exposes 100% of the gRPC API), and for the scale of NVMe/TCP deployments I don't see much benefit from creating amalgamated REST endpoints (and I see the main issue that Dashboard devels will have to take care of the same gRPC-wrapping code in 2 different places).

Reply:
IMO its a good idea to have bulk operations like deleting multiple/adding multiple.
For now, both API and CLI allow user to create their own batch/bash scripts around those which is good practice as far as backend impl is concerned (like fine-grained, declarative and maintainable).

But for UI,if we do same then its no value and we are not providing anything valuable over what CLI has to do.
At least, I wont use the UI then over CLI if its same.

Though its easy for me as dev to maintain such UI and get the task done yet I dont see a benefit for dashboard user.